internal/state/responders.go

```diff
@@ -181,7 +191,22 @@ func (u *fetch) handle(ctx context.Context, tx *ent.Tx, snap *snapshot) ([]respo
 	}
 
 	// Set the new flags as per the fetch response (recent flag is preserved).
-	if err := snap.setMessageFlags(u.messageID, u.flags); err != nil {
+	var newMessageFlags imap.FlagSet
+
+	switch u.fetchFlagOp {
+	case FetchFlagOpAdd:
+		newMessageFlags = curFlags.AddFlagSet(u.flags)
+	case FetchFlagOpRem:
+		newMessageFlags = curFlags.RemoveFlagSet(u.flags)
+	case FetchFlagOpSet:
+		newMessageFlags = u.flags
+	}
+
+	if u.toggleDeleted {
+		newMessageFlags = newMessageFlags.Set(imap.FlagDeleted, curFlags.Contains(imap.FlagDeleted))
+	}
```

Comment:
Is this actually toggling? Shouldn't it be `!curFlags.Contains(imap.FlagDeleted)`?

Reply:
renamed to `cameFromDifferentMailbox` for clarity.